dog_lover/src/breedContext.tsx

```diff
@@ -0,0 +1,135 @@
+};
+export const useVariableOpenContext = () => {
+  return useContext(OpenContext);
+};
+export const useVariableQueryContext = () => {
+  return useContext(QueryContext);
+};
+export const useFunctionHeartContext = () => {
+  return useContext(HeartContextUpdate);
+};
+export const useFunctionFetchBreedsContext = () => {
+  return useContext(FetchBreedsContextUpdate);
+};
+export const useFunctionQueryContext = () => {
+  return useContext(QueryContextUpdate);
+};
+export const useVariableRandomImageContext = () => {
+  return useContext(RandomImageContext);
+};
+export const useFunctionRandomImageContext = () => {
+  return useContext(RandomImageContextUpdate);
+};
+
+export const BreedProvider = ({ children }: { children: any }) => {
+  const [breeds, setBreeds] = useState<DogBreed[]>([]);
+  const [clickedBreed, setClickedBreed] = useState<DogBreedImage[]>([]);
+  const [isOpened, setIsOpened] = useState(false);
+  const [query, setQuery] = useState("");
+  const [randomImage, setRandomImage] = useState<ImageResponse[]>([]);
+
+  const toggle = () => {
+    setIsOpened((wasOpened) => !wasOpened);
+    console.log("toggle");
+  };
+
+  const handleHeartClicked = (breed: string) => () => {
+    setClickedBreed((state) => ({
+      ...state, // <-- copy previous state
+      [breed]: !state[breed], // <-- update value by index key
+    }));
+    toggle();
+  };
+
+  const fetchBreedsList = async () => {
+    // fetch all breeds. Axios requests can be typed for the response
+    try {
+      const {
+        data: { message: allBreeds },
+      } = await axios.get<ListAllResponse>(
+        "https://dog.ceo/api/breeds/list/all"
+      );
+
+      // extract keys from the ListAllResponse.message
+      const keys = Object.keys(allBreeds);
+
+      // resolve images and create DogBreed objects
+      const dogBreeds = await Promise.all(
+        keys.map(async (key) => {
+          const {
+            data: { message: value },
+          } = await axios.get<ImageResponse>(
+            `https://dog.ceo/api/breed/${encodeURIComponent(key)}/images/random`
+          );
+          return { key, value };
+        })
+      );
+      setBreeds(dogBreeds);
+    } catch (err) {
+      console.log(err);
+    }
+  };
+
+  const filterList = (e: any) => {
+    setQuery(e.target.value);
+  };
+
+  const fetchRandomImages = async () => {
+    try {
+      const res: AxiosResponse = await axios.get<ImageResponse>(
+        "https://dog.ceo/api/breeds/image/random"
+      );
+      setRandomImage(res.data);
+    } catch (err) {
+      console.log(err);
+    }
+  };
+
+  return (
+    <ClickedBreedContext.Provider value={clickedBreed}>
+      <BreedContext.Provider value={breeds}>
+        <HeartContextUpdate.Provider value={handleHeartClicked}>
+          <FetchBreedsContextUpdate.Provider value={fetchBreedsList}>
+            <OpenContext.Provider value={isOpened}>
+              <QueryContext.Provider value={query}>
```

Comment:
It seems that you are using each provider for a single purpose. If you had to combine them how would you approach this?

Reply:
So this is my first attempt at using Typescript AND Context API, so when I tried to combine them, which I did at first, I just added all of the variables inside one provider and the functions inside the other, but I was getting errors that I think were mainly because not all of them have the same types. So I think the way to go would be to combine them if they had the same interfaces/types. I am not sure about that one though!